crates/matrix-sdk/src/notification_settings/mod.rs

```diff
@@ -12,3 +26,751 @@ pub enum RoomNotificationMode {
     /// Do not receive any notifications.
     Mute,
 }
+
+/// A high-level API to manage the client owner's push notification settings.
+#[derive(Debug, Clone)]
+pub struct NotificationSettings {
+    /// The underlying HTTP client.
+    client: Client,
+    /// Owner's account push rules. They will be updated on sync.
+    rules: Arc<RwLock<Rules>>,
+    /// Event handler for push rules event
+    push_rules_event_handler: EventHandlerHandle,
+}
+
+impl Drop for NotificationSettings {
+    fn drop(&mut self) {
+        self.client.remove_event_handler(self.push_rules_event_handler.clone());
+    }
+}
+
+impl NotificationSettings {
+    /// Build a new `NotificationSettings``
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - A `Client` used to perform API calls
+    /// * `ruleset` - A `Ruleset` containing account's owner push rules
+    pub fn new(client: Client, ruleset: Ruleset) -> Self {
+        let rules = Arc::new(RwLock::new(Rules::new(ruleset)));
+
+        // Listen for PushRulesEvent
+        let rules_clone = rules.clone();
+        let push_rules_event_handler = client.add_event_handler(move |ev: PushRulesEvent| {
+            let rules = rules_clone.clone();
+            async move {
+                *rules.write().await = Rules::new(ev.content.global);
+            }
+        });
+
+        Self { client, rules, push_rules_event_handler }
+    }
+
+    /// Replace the internal ruleset
+    ///
+    /// # Arguments
+    ///
+    /// * `ruleset` - A `Ruleset` containing account's owner push rules
+    async fn set_ruleset(&self, ruleset: &Ruleset) {
+        *self.rules.write().await = Rules::new(ruleset.clone())
+    }
+
+    /// Get a new `Rules` instance to interact with the ruleset.
+    async fn rules(&self) -> Rules {
```

Comment:
Could this return a `RwLockReadGuard<'_, Rules>`, instead? So it's the caller's responsibility to decide whether to clone or not, depending on its needs (I suspect most of the time it's not required to clone).